src/v2/providers/database.ts

```diff
@@ -0,0 +1,452 @@
+    handler,
+    true
+  ) as CloudFunction<DatabaseEvent<Change<DataSnapshot>>>;
+}
+
+/**
+ * Event handler which triggers when data is deleted in Realtime Database.
+ *
+ * @param reference - The database reference path to trigger on.
+ * @param handler - Event handler which is run every time a Realtime Database deletion occurs.
+ */
+export function onRefDeleted(
+  reference: string,
+  handler: (event: DatabaseEvent<DataSnapshot>) => any | Promise<any>
+): CloudFunction<DatabaseEvent<DataSnapshot>>;
+
+/**
+ * Event handler which triggers when data is deleted in Realtime Database.
+ *
+ * @param opts - Options that can be set on an individual event-handling function.
+ * @param handler - Event handler which is run every time a Realtime Database deletion occurs.
+ */
+export function onRefDeleted(
+  opts: ReferenceOptions,
+  handler: (event: DatabaseEvent<DataSnapshot>) => any | Promise<any>
+): CloudFunction<DatabaseEvent<DataSnapshot>>;
+
+/**
+ * Event handler which triggers when data is deleted in Realtime Database.
+ *
+ * @param referenceOrOpts - Options or a string reference.
+ * @param handler - Event handler which is run every time a Realtime Database deletion occurs.
+ */
+export function onRefDeleted(
+  referenceOrOpts: string | ReferenceOptions,
+  handler: (event: DatabaseEvent<DataSnapshot>) => any | Promise<any>
+): CloudFunction<DatabaseEvent<DataSnapshot>> {
+  return onOperation(
+    deletedEventType,
+    referenceOrOpts,
+    handler,
+    false
+  ) as CloudFunction<DatabaseEvent<DataSnapshot>>;
+}
+
+/** @internal */
+export function getOpts(referenceOrOpts: string | ReferenceOptions) {
+  let path: string, instance: string, opts: options.EventHandlerOptions;
+  if (typeof referenceOrOpts === 'string') {
+    path = normalizePath(referenceOrOpts);
+    instance = '*';
+    opts = {};
+  } else {
+    path = normalizePath(referenceOrOpts.ref);
+    instance = referenceOrOpts.instance || '*';
+    opts = { ...referenceOrOpts };
+    delete (opts as any).ref;
+    delete (opts as any).instance;
+  }
+
+  return {
+    path,
+    instance,
+    opts,
+  };
+}
+
+/** @hidden */
+function trimParam(param: string) {
+  const paramNoBraces = param.slice(1, -1);
+  if (paramNoBraces.includes('=')) {
+    return paramNoBraces.slice(0, paramNoBraces.indexOf('='));
+  }
+  return paramNoBraces;
+}
+
+/** @internal */
+export function matchParams(
+  pathSegments: string[],
+  wildcards: RegExpMatchArray,
+  refSegments: string[],
+  params: Record<string, string>
+): void {
+  let wildcardNdx = 0,
+    refNdx = 0;
+  for (
+    let pathNdx = 0;
+    pathNdx < pathSegments.length && wildcardNdx < wildcards.length;
+    pathNdx++
+  ) {
+    const multiSegmentWildcard = pathSegments[pathNdx].includes('**');
+
+    const remainingPathSegments = pathSegments.length - 1 - pathNdx;
+    const nextRefElement = refSegments.length - remainingPathSegments;
+
+    if (pathSegments[pathNdx] === wildcards[wildcardNdx]) {
```

Comment:
This logic breaks if there are two `**` components (you'll need recursion or maybe dynamic programming). Is that valid according to Eventarc? I strongly assume `**` was only allowed as the last component of a path in One Platform simply because allowing it in the middle allows expensive/ambiguous match calls (e.g. what does `/{a=**}/{b}/{c=**}` match against `/a/b/c/d/e/f`.

As it stands now, you might want to make a proper well-tested class for PathPattern with something like

```
interface PathSegment { ... }
class LiteralSegment extends PathSegment;
class SingleCaptureSegment extends PathSegment;
class GlobCaptureSegment extends PathSegment;
class PathPattern {
  private segments: PathSegment[];
  /** @throws on validation error */
  static compile(raw: string);
  matches(path: string): boolean;
  // If false, we don't need to use pathPattern as our eventarc match type.
  hasCaptures(): boolean;
  extractMatches(path: string): Record<string, string>;
}
```

Reply:
Per the eventarc docs (https://cloud.google.com/eventarc/docs/path-patterns), we can only have at most one mutli segment wildcard `**` and it does not have to be at the end.

Yeah having a separate class makes sense to me, I'll implement it. However, I don't think there should be any validation in the SDK since at runtime the function should always have a valid pattern string from eventarc. I think that validation logic should go into the CLI